Local send fix

gabr1gus opened the pull request:
Simples fix para o problema de transação de send e receive possuirem o mesmo txid
Agora o txid e o tipo fazem parte de um indice composto, pode se ter 2 transaçoes com mesmo txid, mas so pode ter uma combinação de txid-tipo
Remoção dos index antigos diretamente no DB é necessário para alterações surtirem efeito

gabr1gus committed:
criado collection sendTtransaction

--need testing
--preciso confirmar se são apenas esses lugares que salvam transações enviadas

gabr1gus committed:
adicionado index composto em tx

--desfeito o comit anterior
--não testado, mas deve funcionar

gabr1gus committed:
main server transaction schema index

--adicionado o index ao schema de trasaçoes do main server
--removido unique do txid, agora a unicalidade do txid é tratada pelo index

GCastilho commented:
>Remoção dos index antigos diretamente no DB é necessário para alterações surtirem efeito

"diretamente no db" seria no meu database?

gabr1gus commented:
> "diretamente no db" seria no meu database?

sim, mongosse (ou talvez seja o proprio mongo) cria index toda vez que tem um elemento "unique" no schema, o que é OK desde que vc nao mude o schema depois para que esse elemento não seja mais unique, vc pode mudar no schema, mas ainda continua no DB
para corrigir isso da pra fazer algo como " db.runCommand( { dropIndexes: "collection", index: "*" } ) " no shel pra remover todos os index, depois o mongoose cuida de recriar eles, da pra remover manualmente no mongoDB compass tbm, deve dar pra fazer isso no equivalente que você (@GCastilho)  usa

GCastilho committed:
Removido autoIndex

-- Removido autoIndex dos módulos externos, adicionado comment para poder lembrar q ele existe
-- Aumentado tempo do timeout de um request aos módulos externos para 10 segundos

GCastilho committed:
Merge branch 'master' into local-send-fix

gabr1gus committed:
correçoçao para passar no test

--agora txid se comporta como sparse num index composto
--nota: testes são uteis

GCastilho commented on src/db/models/transaction.ts:
```diff
@@ -232,6 +231,9 @@ const TransactionSchema: Schema = new Schema({
  * with the same txid
  *
+ * partialFilterExpression especifica que esse index só existe para os campos em
+ * que um txid esta presente, possibilitando um "sparse" de unico campo em um index composto
```
Isso é excelente, não sabia q existia

GCastilho commented on src/db/models/transaction.ts:
```diff
@@ -229,6 +227,27 @@ const TransactionSchema: Schema = new Schema({
+	type: 1
+}, {
+	unique: true,
+	partialFilterExpression: {
+		txid: {$exists: true},
+	}
```
O external modules tbm não deveria ter isso? Pq se eu n me engano o sistema de salvar transações dele é o mesmo, n é? (also, bota espaços dentro das chaves, assim: `{ $exists: true }`

gabr1gus replied:
eu achava que nao por que lembrava que estava como "required = true" mas nao é o caso é o inverso na verdade, ja resolvo isso

gabr1gus replied:
resolvi isso

gabr1gus committed:
minor change

gabr1gus committed:
minor change

GCastilho commented:
@gabr1gus Tudo pronto ou falta alguma coisa?
Bota ready se tiver pronto q eu vou fazer os testes

GCastilho committed:
Merge branch 'master' into local-send-fix

GCastilho committed:
Adicionado testes

-- Adicionado testes (no main server) para garantir que as transações de send e receive com o mesmo txid serão salvas

GCastilho reviewed (approved):
Complex problems requires simples solutions
